### DISCUSSION TITLE
"core-version" field: what purpose does it serve?

### DISCUSSION THREAD

**Discussion by saqimtiaz** [TiddlyWiki/TiddlyWiki5#5863] [Q&A]
End-user and developer documentation suggests that plugins should have a `core-version` field with examples for values being `>=5.0.8` and `>=5.0.0`

Is this enforced as a requirement during import by drag and drop or when installing via a plugin library? It isn't as far as I can tell but perhaps I am missing something. Neither is this information visible to users while installing plugins via a plugin library or via drag and drop.

Furthermore, attempts to check for core-version compatibility with wikitext are not particularly easy since the `compare[]` operator for type `version` does not support patterns like `>=5.1.0`.

- What purpose does this field serve?
  - Given a `core-version` field with value `5.1.23` is it meant to suggest that this the minimum core version, or that the plugin works with precisely that core version?
- Should we add enforcement or at least visual affordance for the `core-version` field during the plugin installation process?
  - If this field is not enforced as a plugin requirement, this should be clarified in the documentation.
- Should we consider extending the compare operator to parse version strings like `>=5.1.0`?

**Comment by pmario** [comment]
I'm using it as the minimum requirement for my plugins. .. BUT I thought it would be enforced. :/ ... 

>Should we consider extending the compare operator to parse version strings like >=5.1.0?

I think, that would be good.

**Comment by Jermolene** [comment]
Thanks @saqimtiaz @pmario I think you've stumbled on our longest standing bit of cargo culting. I do have a faint memory of writing code to enforce core-version checks on plugins, but evidently it was dropped if it was there at all.

> Is this enforced as a requirement during import by drag and drop or when installing via a plugin library? It isn't as far as I can tell but perhaps I am missing something. Neither is this information visible to users while installing plugins via a plugin library or via drag and drop.

I was slightly surprised to discover that there's no enforcement at the moment.

> Furthermore, attempts to check for core-version compatibility with wikitext are not particularly easy since the `compare[]` operator for type `version` does not support patterns like `>=5.1.0`.

Ouch. The idea was to support the semver comparisons that npm supports (https://github.com/npm/node-semver).

> * What purpose does this field serve?
>   
>   * Given a `core-version` field with value `5.1.23` is it meant to suggest that this the minimum core version, or that the plugin works with precisely that core version?

I think that that would indeed tie the plugin to a particular core version number.

> * Should we add enforcement or at least visual affordance for the `core-version` field during the plugin installation process?

It seems like a good idea, but I do wonder whether it would be worth the effort.

>   * If this field is not enforced as a plugin requirement, this should be clarified in the documentation.

Yes; if we decide not to enforce it, it still seems perfectly reasonable for us to keep it as an advisory field for humans.

> * Should we consider extending the compare operator to parse version strings like `>=5.1.0`?

Hmmm so we'd have something like `[[v5.1.23-prerelease]compare:version[>=v5.1.22]]`, where dropping the "mode" implies that it should be parsed from the operand. Another issue is that the `compare` operator at the moment expects a lowercase "v" which is not used in the "core-version" field.
